lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/DropIfFlaggedFilter.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.analysis.miscellaneous;
+
+import java.io.IOException;
+import java.util.UUID;
+
+import org.apache.lucene.analysis.TokenFilter;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.FlagsAttribute;
+
+/**
+ * Allows Tokens with a given combination of flags to be dropped.
+ *
+ * @see DropIfFlaggedFilterFactory
+ */
+public class DropIfFlaggedFilter extends TokenFilter {
+
+  private int dropFlags;
+
+  private CharTermAttribute attribute = getAttribute(CharTermAttribute.class);
+  private boolean firstToken = true;
+
+  /**
+   * Construct a token stream filtering the given input.
+   *
+   * @param input the source stream
+   * @param dropFlags a combination of flags that indicates that the token should be dropped.
+   */
+  @SuppressWarnings("WeakerAccess")
+  protected DropIfFlaggedFilter(TokenStream input, int dropFlags) {
+    super(input);
+    this.dropFlags = dropFlags;
+  }
+
+  @Override
+  public final boolean incrementToken() throws IOException {
+    boolean result;
+    boolean dropToken;
+    do {
+      result = input.incrementToken();
+      dropToken = (getAttribute(FlagsAttribute.class).getFlags() & dropFlags) == dropFlags;
```

Comment:
you have to move this to final fiel and use FlagsAttribute flagsAttr = addAttribute(FlagsAttribute.class)

Reply:
I guess this has been working because org.apache.lucene.analysis.Token happens to have this attribute already, but of course that's not guaranteed to be the impl one gets.

The final field... I've seen this done in a number of places including doc examples but I don't recall (and haven't been able to locate documentation as to) why it's important. Is there anything to it beyond saving an Class.cast() and a Map.get() call? It is a very hot area of the code so maybe something that small does matter here... Also I wonder if anything like the following has been considered so that users don't have to explicitly add attributes:

```
  public final <T extends Attribute> T getAttributeSafe(Class<T> attClass) {
    return attClass.cast(attributes.computeIfAbsent(attClass, this::addAttribute));
  }
```
That doesn't quite work with the current generics and obviously not for this ticket but is that an already discarded idea?